multi-era/rust/src/utils.rs

```diff
@@ -60,6 +98,1204 @@ impl MultiEraBlock {
+                .transaction_witness_sets
+                .iter()
+                .map(|wits| wits.clone().into())
+                .collect(),
+            Self::Alonzo(block) => block
+                .transaction_witness_sets
+                .iter()
+                .map(|wits| wits.clone().into())
+                .collect(),
+            Self::Babbage(block) => block
+                .transaction_witness_sets
+                .iter()
+                .map(|wits| wits.clone().into())
+                .collect(),
+            Self::Conway(block) => block.transaction_witness_sets.clone(),
+        }
+    }
+
+    pub fn auxiliary_data_set(&self) -> OrderedHashMap<TransactionIndex, AuxiliaryData> {
+        match self {
+            Self::Byron(_block) => OrderedHashMap::default(),
+            Self::Shelley(block) => block
+                .transaction_metadata_set
+                .iter()
+                .map(|(i, md)| (*i, AuxiliaryData::new_shelley(md.clone())))
+                .collect(),
+            Self::Allegra(block) => block
+                .auxiliary_data_set
+                .iter()
+                .map(|(i, md)| (*i, md.clone().into()))
+                .collect(),
+            Self::Mary(block) => block
+                .auxiliary_data_set
+                .iter()
+                .map(|(i, md)| (*i, md.clone().into()))
+                .collect(),
+            Self::Alonzo(block) => block
+                .auxiliary_data_set
+                .iter()
+                .map(|(i, md)| (*i, md.clone().into()))
+                .collect(),
+            Self::Babbage(block) => block
+                .auxiliary_data_set
+                .iter()
+                .map(|(i, md)| (*i, md.clone().into()))
+                .collect(),
+            Self::Conway(block) => block.auxiliary_data_set.clone(),
+        }
+    }
+
+    pub fn invalid_transactions(&self) -> Vec<TransactionIndex> {
+        match self {
+            Self::Byron(_block) => vec![],
+            Self::Shelley(_block) => vec![],
+            Self::Allegra(_block) => vec![],
+            Self::Mary(_block) => vec![],
+            Self::Alonzo(block) => block.invalid_transactions.clone(),
+            Self::Babbage(block) => block.invalid_transactions.clone(),
+            Self::Conway(block) => block.invalid_transactions.clone(),
+        }
+    }
+}
+
+#[derive(Clone, Debug, serde::Deserialize, serde::Serialize, schemars::JsonSchema)]
+pub enum MultiEraBlockHeader {
+    ByronEB(EbbHead),
+    Byron(ByronBlockHeader),
+    Shelley(ShelleyHeader),
+    Babbage(Header),
+}
+
+impl MultiEraBlockHeader {
+    pub fn block_number(&self) -> Option<u64> {
+        match self {
+            Self::ByronEB(_) => None,
+            Self::Byron(_) => None,
+            Self::Shelley(header) => Some(header.body.block_number),
+            Self::Babbage(header) => Some(header.header_body.block_number),
+        }
+    }
+
+    pub fn slot(&self) -> Option<u64> {
```

Comment:
Do we want to do something for Byron? I'm not sure how the transition went there. IIRC it used to be like 22600 or something slots per epoch, and the slots reset but I don't really know/remember. Should we calculate it like that? Or is that only on mainnet that it was that number? Or should we just only have this work on shelley-onward blocks?

Reply:
need to check